### PR TITLE
Return a HTTP 409 for conflicts when creating add-ons/versions through the API

### LINKED ISSUE #20387: Return HTTP 409 Conflict error in add-on API when submitting a duplicate version
In the new add-on API, submitting a version that already exists returns a 400 error with `Version {version_string} was uploaded before and deleted.` or `Version {version_string} already exists.`.

It should return a 409 instead to make it easier for clients to programmatically differentiate between a duplicate submission and more general bad request errors.

### PR DESCRIPTION
**Pull request by diox**
Also add a few more add-on creation tests

Fixes #20387

### CODE REVIEW

File: src/olympia/addons/serializers.py [hunk 1]
```diff
@@ -486,7 +487,10 @@ def validate_upload(self, value):
         if not own_upload or not value.valid or value.validation_timeout:
             raise exceptions.ValidationError(gettext('Upload is not valid.'))
         # Parse the file to get and validate package data with the addon.
-        self.parsed_data = parse_addon(value, addon=self.addon, user=request.user)
+        try:
+            self.parsed_data = parse_addon(value, addon=self.addon, user=request.user)
+        except DuplicateAddonID as exc:
+            raise Conflict({self.field_name: exc.messages})
```
Comment (eviljeff) [inline src/olympia/addons/serializers.py:493]: I guess you need to manually wrap the `exc.message` in a dict here (exceptions from a `validate_xxx` function should already get wrapped as `{xxx: exception}`) because an `APIException` subclass is being raised rather than a `ValidationError`?

File: src/olympia/addons/tests/test_views.py [hunk 2]
```diff
@@ -3395,8 +3490,8 @@ def test_duplicate_version_number_error(self):
             self.url,
             data=self.minimal_data,
         )
-        assert response.status_code == 400, response.content
-        assert response.data == {
+        assert response.status_code == 409, response.content
+        assert response.json() == {
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_views.py:3494]: out of interest, why the change to `response.json()` from `response.data`? `response.data` seems to be - by far- the most common use in our tests.

**Review by eviljeff — APPROVED** [review]
One drawback of this change is if any api consumer is directly checking for `400` for validation errors their code will now break... but agree it's an overall good idea.

File: src/olympia/addons/tests/test_views.py (hunk 2 above)
Comment (diox) [reply src/olympia/addons/tests/test_views.py:3494]: `response.data` is the python representation of what _would_ be rendered, but can still contain objects like `ErrorDetail`. `response.json()` is the rendered json converted back to Python. The latter is easier to compare to deep dicts.

File: src/olympia/addons/serializers.py (hunk 1 above)
Comment (diox) [reply src/olympia/addons/serializers.py:493]: Yes. That is annoying, but there is no method to call to do that, the `ValidationError` conversion happens in DRF `is_valid()` code (and there is something else for authentication errors or 500 exceptions).

I have raised a non-dict, but then it wouldn't have been consistent with the way errors are typically returned.
